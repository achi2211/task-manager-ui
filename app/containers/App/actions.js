/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_TASKS,
  LOAD_TASKS_SUCCESS,
  LOAD_TASKS_ERROR,
  UPDATE_TASK,
  UPDATE_TASK_SUCCESS,
  UPDATE_TASK_ERROR,
} from './constants';

/**
 * Load the tasks, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_TASKS
 */
export function loadTasks() {
  return {
    type: LOAD_TASKS,
  };
}

/**
 * Dispatched when the tasks are loaded by the request saga
 *
 * @param  {array} tasks The
 * @param  {int} limit The current limit
 *
 * @return {object}      An action object with a type of LOAD_TASKS_SUCCESS passing the tasks
 */
export function tasksLoaded(tasks, limit) {
  return {
    type: LOAD_TASKS_SUCCESS,
    tasks,
    limit,
  };
}

/**
 * Dispatched when loading the tasks fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_TASKS_ERROR passing the error
 */
export function taskLoadingError(error) {
  return {
    type: LOAD_TASKS_ERROR,
    error,
  };
}

/**
 * update the tasks, this action starts the request saga
 *
 * @return {object} An action object with a type of TASK
 */
export function updateTask(taskId) {
  return {
    type: UPDATE_TASK,
    taskId
  };
}

/**
* Dispatched when the task are updated by the request saga
*
* @return {object} An action object with a type of LOAD_TASKS_SUCCESS passing the tasks
*/
export function taskUpdated() {
 return {
   type: UPDATE_TASK_SUCCESS,
 };
}

/**
 * Dispatched when loading the tasks fails
 *
 * @param  {object} error The error
 *
 * @return {object} An action object with a type of LOAD_TASKS_ERROR passing the error
 */
export function taskUpdatedError(error) {
  return {
    type: UPDATE_TASK_ERROR,
    error,
  };
}
