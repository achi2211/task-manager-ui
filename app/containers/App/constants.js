/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const LOAD_TASKS = 'boilerplate/App/LOAD_TASKS';
export const LOAD_TASKS_SUCCESS = 'boilerplate/App/LOAD_TASKS_SUCCESS';
export const LOAD_TASKS_ERROR = 'boilerplate/App/LOAD_TASKS_ERROR';

export const UPDATE_TASK = 'boilerplate/App/UPDATE_TASK';
export const UPDATE_TASK_SUCCESS = 'boilerplate/App/UPDATE_TASK_SUCCESS';
export const UPDATE_TASK_ERROR = 'boilerplate/App/UPDATE_TASK_ERROR';
