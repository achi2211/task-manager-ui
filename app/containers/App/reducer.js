/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import {
  LOAD_TASKS_SUCCESS,
  LOAD_TASKS,
  LOAD_TASKS_ERROR,
  UPDATE_TASK,
  UPDATE_TASK_SUCCESS,
  UPDATE_TASK_ERROR,
} from './constants';

// The initial state of the App
export const initialState = {
  loading: false,
  error: false,
  currentUser: false,
  userData: {
    tasks: false,
  },
  task_update: false,
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case LOAD_TASKS:
        draft.loading = true;
        draft.error = false;
        draft.userData.tasks = false;
        break;

      case LOAD_TASKS_SUCCESS:
        draft.userData.tasks = action.tasks;
        draft.loading = false;
        draft.currentUser = action.taskLimit;
        break;

      case LOAD_TASKS_ERROR:
        draft.error = action.error;
        draft.loading = false;
        break;

      case UPDATE_TASK:
        draft.error = false;
        draft.task_update = false;
        break;

      case UPDATE_TASK_SUCCESS:
        draft.task_update = true;
        draft.error = false;
        draft.loading = false;
        break;

      case UPDATE_TASK_ERROR:
        draft.error = action.error;
        draft.loading = false;
        break;
    }
  });

export default appReducer;
