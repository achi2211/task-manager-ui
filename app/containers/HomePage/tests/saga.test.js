/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { LOAD_TASKS } from 'containers/App/constants';
import { tasksLoaded, taskLoadingError } from 'containers/App/actions';

import taskListData, { getTasks } from '../saga';

const taskLimit = 'mxstbr';

/* eslint-disable redux-saga/yield-effects */
describe('getTasks Saga', () => {
  let getReposGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getReposGenerator = getTasks();

    const selectDescriptor = getReposGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getReposGenerator.next(taskLimit).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the tasksLoaded action if it requests the data successfully', () => {
    const response = [
      {
        name: 'First repo',
      },
      {
        name: 'Second repo',
      },
    ];
    const putDescriptor = getReposGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(tasksLoaded(response, taskLimit)));
  });

  it('should call the taskLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getReposGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(taskLoadingError(response)));
  });
});

describe('githubDataSaga Saga', () => {
  const githubDataSaga = taskListData();

  it('should start task to watch for LOAD_TASKS action', () => {
    const takeLatestDescriptor = githubDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_TASKS, getTasks));
  });
});
