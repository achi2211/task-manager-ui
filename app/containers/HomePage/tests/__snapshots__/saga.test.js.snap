// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTasks Saga should call the taskLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getTasks Saga should call the taskLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr/tasks?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getTasks Saga should dispatch the tasksLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getTasks Saga should dispatch the tasksLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://api.github.com/users/mxstbr/tasks?type=all&sort=updated",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
